src/redis_info_provider/local_watcher.py

```diff
@@ -59,20 +69,24 @@ def __exit__(self, exception_type, exception_value, traceback):
     def _greenlet_main(self):
         # type: () -> None
 
-        while not self._should_stop.is_set():
-            published_shard_ids = ShardPublisher.get_live_shard_ids()
-            live_shards = self._get_live_shards()
-            live_shard_ids = set(live_shards.keys())
-
-            logger.info('Updated Redis shards: %s', live_shard_ids)
-
-            for shard_id in live_shard_ids - published_shard_ids:
-                # New shard
-                ShardPublisher.add_shard(live_shards[shard_id])
-            for shard_id in published_shard_ids - live_shard_ids:
-                # Removed shard
-                ShardPublisher.del_shard(shard_id)
-            gevent.sleep(self._update_freq)
+        try:
+            while not self._should_stop.is_set():
+                published_shard_ids = ShardPublisher.get_live_shard_ids()
+                live_shards = self._get_live_shards()
+                live_shard_ids = set(live_shards.keys())
+
+                logger.info('Updated Redis shards: %s', live_shard_ids)
+
+                for shard_id in live_shard_ids - published_shard_ids:
+                    # New shard
+                    ShardPublisher.add_shard(live_shards[shard_id])
+                for shard_id in published_shard_ids - live_shard_ids:
+                    # Removed shard
+                    ShardPublisher.del_shard(shard_id)
+                gevent.sleep(self._update_freq)
+        except Exception as e:
```

Comment:
Correct me if I'm wrong, but won't this also `set_exception` when we call `InfoPoller.stop` (since IIUC that causes `gevent.GreenletExit` to be raised inside the greenlets)? If so maybe we want to exclude that one from the handling?